CONTRIBUTING.md

```diff
@@ -481,6 +482,41 @@ run as part of the regular build, since they have some particular requirements.
 See the [CLI CONTRIBUTING.md file](packages/aws-cdk/CONTRIBUTING.md) for
 more information on running those tests.
 
+### Future Flags
+
+Sometimes we want to introduce new breaking behavior because we believe this is the correct default behavior for the CDK. The problem of course is that breaking changes
+are only allowed in major versions and those are rare.
+
+To address this need, we have a pattern/mechanism called "future flags". It allows us to introduce new breaking behavior which is disabled by default (so existing projects will not be affected) but enabled automatically for new projects created through `cdk init`.
```

Comment:
After a while, this is going to lead to an overgrowth of flags in `cdk.json`, all of which need to be set, AND we will only get people on new behavior if they use `cdk init`, not any other way.

I was thinking that a combination of straight-up feature flags and a variant of Romain's suggestion would be more effective in getting people onto new behavior as soon as possible, with minimum friction, while not breaking existing code.

We still have feature flags, and users can still toggle them on or off. However, the *default* on a fresh project would be the *new* behavior instead of the *old* behavior, without the need to explicitly turn it on.

How do we make sure behavior of existing code doesn't change? We derive the default value of the flag using the version of the CDK used to create the project (storing a "behavior" version in `cdk.json`). So:

```
# Everything as it was for release 1.16 (no duplicate stack names)
{
  "featureDefaults": "1.16.0",
}

# Everything as it was for release 1.16 but turn on duplicate stack names
{
  "featureDefaults": "1.16.0",
  "duplicateStackNames": true,
}

# Everything as it was for release 1.17 (enables duplicate stack names)
{
  "featureDefaults": "1.17.0"
}

# Everything as it was for release 1.17 but disable duplicate stack names
{
  "featureDefaults": "1.17.0",
  "duplicatStackNames": false
}
```

Totally willing to have a discussion afterwards on how to encode these flags and their versions and defaults in code in a way that makes it as unlikely as possible that programmers will screw something about this up (which I think is totally feasible), but I'm leaving that out of scope until we agree on the user experience.

Reply:
Let's continue this thread over #4925